Src/Newtonsoft.Json/JsonWriter.Async.cs

```diff
@@ -0,0 +1,1690 @@
+            }
+
+            _currentState = newState;
+
+            if (_formatting == Formatting.Indented)
+            {
+                switch (oldState)
+                {
+                    case State.Start:
+                        break;
+                    case State.Property:
+                        return tokenBeingWritten == JsonToken.PropertyName ? AutoCompleteAsync(tokenBeingWritten, oldState, cancellationToken) : WriteIndentSpaceAsync(cancellationToken);
+                    case State.ArrayStart:
+                    case State.ConstructorStart:
+                        return WriteIndentAsync(cancellationToken);
+                    case State.Array:
+                    case State.Constructor:
+                        return tokenBeingWritten == JsonToken.Comment ? WriteIndentAsync(cancellationToken) : AutoCompleteAsync(tokenBeingWritten, oldState, cancellationToken);
+                    case State.Object:
+                        switch (tokenBeingWritten)
+                        {
+                            case JsonToken.Comment:
+                                break;
+                            case JsonToken.PropertyName:
+                                return AutoCompleteAsync(tokenBeingWritten, oldState, cancellationToken);
+                            default:
+                                return WriteValueDelimiterAsync(cancellationToken);
+                        }
+
+                        break;
+                    default:
+                        if (tokenBeingWritten == JsonToken.PropertyName)
+                        {
+                            return WriteIndentAsync(cancellationToken);
+                        }
+
+                        break;
+                }
+            }
+            else if (tokenBeingWritten != JsonToken.Comment)
+            {
+                switch (oldState)
+                {
+                    case State.Object:
+                    case State.Array:
+                    case State.Constructor:
+                        return WriteValueDelimiterAsync(cancellationToken);
+                }
+            }
+
+            return AsyncUtils.CompletedTask;
+        }
+
+        private async Task AutoCompleteAsync(JsonToken tokenBeingWritten, State oldState, CancellationToken cancellationToken)
+        {
+            Debug.Assert(_formatting == Formatting.Indented);
+            if ((oldState == State.Object || oldState == State.Array || oldState == State.Constructor) && tokenBeingWritten != JsonToken.Comment)
+            {
+                await WriteValueDelimiterAsync(cancellationToken).ConfigureAwait(false);
+            }
+
+            if (oldState == State.Property)
+            {
+                await WriteIndentSpaceAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
I think this path isn't reachable, but I'm not 100% sure. Cutting it would be a slight improvement, but only slight since once you've one `await` having even a lot more is just adding states to the `MoveNext()` and so not a big further impact.